### ISSUE TITLE
Java Indexer: optionally add childof edges from all anchors to their parent scope [phab:D2217]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2681]
_Created by **schroederc** at 2018-03-14 17:58:51_: 

This is a generalization of callsite anchors, allowing for the same hierarchical queries that make the Kythe callgraph possible.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2018-03-14 18:10:48_:

s/semantic/lexical/ ?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2018-03-14 18:10:48_:

Does `param` have a scope we should be testing for?

Shouldn't most of these assertions be true without the flag introduced in this CL? I don't see any tests of childof edges for non-call references, just defines/bindings.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2018-03-14 18:15:07_:

Right now the parameter does not have a parent.

These `childof` edges are from the anchors, not the semantic nodes they reference.  These assertions will fail without the `--emit_anchor_scopes` feature.

**Comment by kythe** [comment]
_**schroederc** wrote at 2018-03-14 18:15:15_:

Fix comment

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2018-03-14 18:45:55_:

Ah, fair enough, thanks.

In that case these edges seem mostly redundant with the semantic childof edges?

Given the potentially catastrophic increase in data these could produce, it might make more sense to have the flag be somewhat more granular, e.g

--emit_anchor_scopes={CALL_REF, NON_LOCAL_REF, LOCAL_REF, ALL_REF, BINDING, ALL} 

or something?

**Comment by kythe** [comment]
_**schroederc** wrote at 2018-03-14 19:32:47_:

Additional tests and cleanup

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2018-03-14 19:33:11_:

The ones for definitions are redundant because the following relation already exists:

```
anchor -[defines/binding]> semantic -[childof]> semantic_scope
```

We're just adding a shortcut edge between the anchor and the scope in this case.

However, that relation doesn't help with references since the semantic scope won't match the lexical context of the anchor, and references are the much more common case.

Adding the more granular approach will take a bit more work.  I was just planning on running the full experiment to see the overall damage.  When it becomes necessary, I will come back to try other approaches.

**Comment by kythe** [comment]
_**schroederc** wrote at 2018-03-15 16:45:27_:

ping

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2018-03-15 16:48:03_

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2018-03-15 16:51:44_:

Fair enough, can we add some verifier assertions for plain ol' ref nodes, regardless?

**Comment by kythe** [comment]
_**shahms** required changes at 2018-03-15 16:51:44_

**Comment by kythe** [comment]
_**shahms** wrote at 2018-03-15 16:52:09_:

Oh, you did, thanks!

**Comment by kythe** [comment]
_**shahms** accepted this revision at 2018-03-15 16:52:09_

**Comment by kythe** [comment]
_**schroederc** closed via commit 0a3afc9bffba3e09c638ae28575605a1441ebf5e at 2018-03-15 16:52:54_
